src/computedevs.jl

```diff
@@ -0,0 +1,80 @@
+# adaptors for converting abstract arrays to Base.Array
+
+
+"""
+    abstract type AbstractComputingDevice
+
+Supertype for CPU and GPU computing devices.
+
+`adapt(dev::AbstractComputingDevice, x)` adapts `x` for `dev`.
+"""
+abstract type AbstractComputingDevice end
+export AbstractComputingDevice
+
+
+"""
+    struct ComputingDeviceIndependent <: AbstractComputingDevice
+
+`get_computing_device(x) === ComputingDeviceIndependent()` indicates
+that `x` is not tied to a specific computing device. This typically
+means that x is a statically allocated object.
+"""
+struct ComputingDeviceIndependent <: AbstractComputingDevice end
+export ComputingDeviceIndependent
+
+
+"""
+    struct CPUDevice <: AbstractComputingDevice
+
+`CPUDevice()` is the default CPU device.
+"""
+struct CPUDevice <: AbstractComputingDevice end
+export CPUDevice
+
+
+
+"""
+    abstract type AbstractGPUDevice <: AbstractComputingDevice
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractGPUDevice <: AbstractComputingDevice end
```

Comment:
Maybe have an abstract accelerator device, so it covers all non-CPU like TPU, IPU, etc.

Reply:
> Maybe have an abstract accelerator device, so it covers all non-CPU like TPU, IPU, etc.

Yes, I had hope that this might extend to other compute devices types as well one day (FPGA, ...). Do we need to have an `AbtractComputingAccelerator` or `AbstractNonCPUDevice`? Or should difference kinds of non-CPU devices be direct subtypes of `AbstractComputingDevice`, on the same level as `CPUDevice` (so one would check `!(dev isa CPUDevice)`)?

Reply:
Thanks @AriMKatz , good idea! We definitely want to make sure this is done in a way that everyone can utilize.

Reply:
> Maybe instead of ComputingDevice we call it ComputeUnit? We are moving towards heterogeneous system in general and they might not be separated devices

I'm renaming it accordingly.

Reply:
> Maybe have an abstract accelerator device

I'm adding AbstractComputeAccelerator into the type hierarchy.